Model/Indexer/Product/Indexer.php

```diff
@@ -57,14 +57,21 @@ public function __construct(
     public function executeFull()
     {
         if ($this->dataHelper->isFullReindexEnabled()) {
-            // Fetch all enabled products
-            $searchCriteria = $this->searchCriteriaBuilder
-                ->addFilter('status', Status::STATUS_ENABLED, 'eq')
-                ->setPageSize(self::HARD_LIMIT_FOR_PRODUCTS)
-                ->setCurrentPage(1)
-                ->create();
-            $products = $this->productRepository->getList($searchCriteria);
-            $this->productService->update($products->getItems());
+            $pageNumber = 0;
+            $totalItems = 0;
+            do {
+                $pageNumber++;
+
+                $searchCriteria = $this->searchCriteriaBuilder
+                    ->addFilter('status', Status::STATUS_ENABLED, 'eq')
+                    ->setPageSize(self::BATCH_SIZE)
+                    ->setCurrentPage($pageNumber)
+                    ->create();
+                $products = $this->productRepository->getList($searchCriteria);
+                $this->productService->update($products->getItems());
+                $totalItems = $products->getTotalCount();
```

Comment:
getTotalCount may return the total number  of products in database. (not the number of products fetched this time). I am not 100% sure, we should at least check it